### PR TITLE
Unit Test Refactor and Fixes

### LINKED ISSUE #405: [Testing] Refactor all unit tests to address test code duplication
As a developer/user, I want to refactor all unit test files flagged by Sonarqube. Parametrization can be used for this refactoring.

**Description** 
Sonarqube flagged multiple unit test with code duplication. During the presentation, the teacher raised the point that it can be refactored using parametrization. 

**Additional Notes**
Here is a screenshot, highlighted with red for the duplicated portions:
image

**Tasks**
- [x] Refactor all unit tests with code duplication (flagged by Sonarqube)
- [x] Document this refactoring in the release doc after the PR has been merged. Include the PR and the correct Fowler's method (Parametrize function) 
https://refactoring.com/catalog/parameterizeFunction.html

**Acceptance Criteria**
All test methods with code duplication have been refactored with Parametrization.
The refactoring is properly documented in the Release doc.
All unit tests must pass after this refactoring.

### LINKED ISSUE #459: Tests Not Always Using Existing Connection
**Describe the bug**
It appears that tests may not use an existing connection if one is present but are always attempting to create a new connection. If a connection already exists, this can cause an error, resulting in a test not executing. In the backend proper, there is currently a set up in the loadstartup such that an existing connection should be used if possible but tests do not utilize this file.

**To Reproduce**
1. Open a terminal in the Server folder
2. Enter "npm run test"
3. Potentially see error

**Expected behavior**
All tests should run without an error.

**Additional context**
Error message when this occurs: 
> ● Data Set Controller  › Invalid search, no search query entered

    AlreadyHasActiveConnectionError: Cannot create a new connection named "default", because connection with such name already exist and it now has an active connection session.

      at new AlreadyHasActiveConnectionError (src/error/AlreadyHasActiveConnectionError.ts:8:9)
      at ConnectionManager.Object.<anonymous>.ConnectionManager.create (src/connection/ConnectionManager.ts:57:23)
      at Object.<anonymous> (src/index.ts:230:35)
      at step (node_modules/typeorm/node_modules/tslib/tslib.js:141:27)
      at Object.next (node_modules/typeorm/node_modules/tslib/tslib.js:122:57)
      at fulfilled (node_modules/typeorm/node_modules/tslib/tslib.js:112:62)

### PR DESCRIPTION
**Pull request by silent-po2**
This PR fixes the various duplications picked up by sonarqube (#405), fixes the bug with tests not always using an existing connection (#459), adds a unit test for better coverage, and fixes a method name that had a typo.

### CODE REVIEW

File: Server/src/tests/controllers/AdminManagementController.spec.ts [hunk 1]
```diff
@@ -8,7 +8,14 @@ describe('Fetch All Categories Materials Controller ', () => {
     let controller: AdminManagementController;
 
     beforeEach(async () => {
-        await createConnection();
+        try {
+            await createConnection();
```
Comment (ds161833) [inline Server/src/tests/controllers/AdminManagementController.spec.ts:12]: isnt there createOrGetConnection?

File: Server/src/tests/controllers/AdminManagementController.spec.ts [hunk 2]
```diff
@@ -129,9 +120,12 @@ describe('Fetch All Categories Materials Controller ', () => {
                 operation: 'remove'
             }
         }
-        await controller.createPermissionUpdateRequest(mockRequest as Request, mockResponse as Response)
-        expect(mockResponse.json).toBeCalledWith("No such user exists");
-        expect(mockResponse.status).toBeCalledWith(404);
+        await updatePermission(mockRequest as Request, mockResponse as Response, 404, "No such user exists")
     });
 
+    async function updatePermission(mockRequest: Request, mockResponse: Response, status: number, expectedResponse: any) {
```
Comment (ds161833) [inline Server/src/tests/controllers/AdminManagementController.spec.ts]: updatePermissionWithAssert (or expect) would better show what this method does, at first i saw the code and though you just removed the expect statements.

File: Server/src/tests/controllers/CategoryController.spec.ts [hunk 3]
```diff
@@ -9,7 +9,14 @@ describe('Category Controller', () => {
   let controller: CategoryController;
 
   beforeEach(async () => {
-    await createConnection();
+    try {
+      await createConnection();
+    } catch (error) {
+      // If AlreadyHasActiveConnectionError occurs, return already existent connection
+      if (error.name === "AlreadyHasActiveConnectionError") {
+        return getConnectionManager().get();
```
Comment (ds161833) [inline Server/src/tests/controllers/CategoryController.spec.ts:17]: what's the point of returning a getConnectionManager().get() if that is not used, just have an empty catch

File: Server/src/tests/controllers/CategoryController.spec.ts [hunk 4]
```diff
@@ -33,98 +40,95 @@ describe('Category Controller', () => {
     mockRequest = {
       body: validUploadData
     }
-    await controller.createCategory(mockRequest as Request, mockResponse as Response)
+    await createCategory(mockRequest as Request, mockResponse as Response, 201)
```
Comment (ds161833) [inline Server/src/tests/controllers/CategoryController.spec.ts]: same here, createCategory only implies that a category will be created, doesnt imply also there will be expect statements

File: Server/src/tests/services/fetchAllCategoriesMaterialsService.spec.ts [hunk 5]
```diff
@@ -1,22 +1,29 @@
...
+  let retrieveBasicDataObject: fetchAllCategoriesMaterialsService;
+  jest.setTimeout(60000)
 
-    beforeAll(async () => {
-        await createConnection();
-        retrieveBasicDataObject = new fetchAllCategoriesMaterialsService();
-    });
+  beforeAll(async () => {
+    try {
+      await createConnection();
```
Comment (ds161833) [inline Server/src/tests/services/fetchAllCategoriesMaterialsService.spec.ts:10]: in the previous tests you create connection beforeEach, creating it before all sounds like a good idea

**Review by ds161833 — APPROVED** [review]

File: Server/src/tests/controllers/CategoryController.spec.ts (hunk 3 above)
Comment (silent-po2) [reply Server/src/tests/controllers/CategoryController.spec.ts:17]: It does get used. This is the exact same set up we have in our loadStartup

File: Server/src/tests/controllers/AdminManagementController.spec.ts (hunk 1 above)
Comment (silent-po2) [reply Server/src/tests/controllers/AdminManagementController.spec.ts:12]: I do not believe so; looking at the source file of getConnectionManager, getConnection, and createConnection there does not appear to be such a function.

File: Server/src/tests/controllers/AdminManagementController.spec.ts (hunk 2 above)
Comment (silent-po2) [reply Server/src/tests/controllers/AdminManagementController.spec.ts]: Will add

File: Server/src/tests/services/fetchAllCategoriesMaterialsService.spec.ts (hunk 5 above)
Comment (silent-po2) [reply Server/src/tests/services/fetchAllCategoriesMaterialsService.spec.ts:10]: It's true, it isn't unified across tests now that you mention it

File: Server/src/tests/controllers/DatasetController.spec.ts [hunk 6]
```diff
@@ -9,7 +9,14 @@ describe('Data Set Controller ', () => {
     let GetDataControllerController: DataSetController;
 
     beforeEach(async () => {
-        await createConnection();
+        try {
```
Comment (GD-Dheer) [inline Server/src/tests/controllers/DatasetController.spec.ts:12]: In favour of using beforeAll()

**Review by GD-Dheer — APPROVED** [review]
Concerns are already raised! Nice work condensing all our test files

File: Server/src/tests/controllers/DatasetController.spec.ts (hunk 6 above)
Comment (silent-po2) [reply Server/src/tests/controllers/DatasetController.spec.ts:12]: Done in latest commit

**Pull request merged**